packages/arb-avm-cpp/data_storage/src/value/code.cpp

```diff
@@ -0,0 +1,266 @@
+        auto key_vec = segment_key(item.first);
+        auto key = vecToSlice(key_vec);
+        deleteRefCountedData(*transaction.transaction, key, item.second);
+    }
+}
+
+uint64_t getNextSegmentID(const Transaction& transaction) {
+    std::string segment_id_raw;
+    auto s = transaction.transaction->Get(rocksdb::ReadOptions(),
+                                          rocksdb::Slice(max_code_segment_key),
+                                          &segment_id_raw);
+    if (s.IsNotFound()) {
+        return 0;
+    }
+    if (!s.ok()) {
+        throw std::runtime_error("couldn't load segment id");
+    }
+    auto ptr = segment_id_raw.data();
+    return checkpoint::utils::deserialize_uint64(ptr);
+}
+
+void saveCode(Transaction& transaction,
+              const Code& code,
+              std::map<uint64_t, uint64_t>& segment_counts) {
+    auto snapshots = code.snapshot();
+
+    std::vector<unsigned char> value_data;
+    marshal_uint64_t(snapshots.op_count, value_data);
+    auto value_slice = vecToSlice(value_data);
+    auto status = transaction.transaction->Put(
+        rocksdb::Slice(max_code_segment_key), value_slice);
+    if (!status.ok()) {
+        throw std::runtime_error("failed to size mac code segment");
+    }
+
+    std::unordered_map<uint64_t, uint64_t> total_segment_counts;
+    auto current_segment_counts = segment_counts;
+    std::unordered_map<uint64_t, std::vector<unsigned char>>
+        code_segments_to_save;
+    bool saved_segment = true;
+
+    while (saved_segment) {
+        saved_segment = false;
+        std::map<uint64_t, uint64_t> next_segment_counts;
+        for (auto it = current_segment_counts.rbegin();
+             it != current_segment_counts.rend(); ++it) {
+            uint64_t segment_id = it->first;
+            uint64_t reference_count = it->second;
+            total_segment_counts[segment_id] += reference_count;
+
+            if (reference_count == 0) {
+                // If there are no references, don't bother saving
+                continue;
+            }
+            if (code_segments_to_save.find(segment_id) !=
+                code_segments_to_save.end()) {
+                // If we've already saved this segment, there's nothing to do
+                continue;
+            }
+            code_segments_to_save[segment_id] = prepareToSaveCodeSegment(
+                transaction, snapshots.segments[segment_id],
+                next_segment_counts);
+            saved_segment = true;
+        }
+        current_segment_counts = std::move(next_segment_counts);
+    }
+
+    // Now that we've handled all references, save all the serialized segments
+    for (const auto& item : code_segments_to_save) {
+        auto key_vec = segment_key(item.first);
+        auto key = vecToSlice(key_vec);
+        saveRefCountedData(*transaction.transaction, key, item.second,
+                           total_segment_counts[item.first], true);
+    }
+}
```

Comment:
i think this class can be refactored to be less complicated, takes awhile to understand whats going on since theres lots of logic happening without much separation

Reply:
It' significantly improved now, though it's definitely still complicated.